tutorials/src/bin/basic-tutorial-8.rs

```diff
@@ -0,0 +1,262 @@
+        .add_many(
+            &[
+                &appsrc,
+                &tee,
+                &audio_queue,
+                &audio_convert1,
+                &audio_resample,
+                &audio_sink,
+                &video_queue,
+                &audio_convert2,
+                &visual,
+                &video_convert,
+                &video_sink,
+                &app_queue,
+                &appsink,
+            ],
+        )
+        .unwrap();
+
+    gst::Element::link_many(&[&appsrc, &tee]).unwrap();
+    gst::Element::link_many(
+        &[&audio_queue, &audio_convert1, &audio_resample, &audio_sink],
+    ).unwrap();
+    gst::Element::link_many(
+        &[
+            &video_queue,
+            &audio_convert2,
+            &visual,
+            &video_convert,
+            &video_sink,
+        ],
+    ).unwrap();
+    gst::Element::link_many(&[&app_queue, &appsink]).unwrap();
+
+    let tee_audio_pad = tee.get_request_pad("src_%u").unwrap();
+    println!(
+        "Obtained request pad {} for audio branch",
+        tee_audio_pad.get_name()
+    );
+    let queue_audio_pad = audio_queue.get_static_pad("sink").unwrap();
+    tee_audio_pad.link(&queue_audio_pad).into_result().unwrap();
+
+    let tee_video_pad = tee.get_request_pad("src_%u").unwrap();
+    println!(
+        "Obtained request pad {} for video branch",
+        tee_video_pad.get_name()
+    );
+    let queue_video_pad = video_queue.get_static_pad("sink").unwrap();
+    tee_video_pad.link(&queue_video_pad).into_result().unwrap();
+    let tee_app_pad = tee.get_request_pad("src_%u").unwrap();
+    let queue_app_pad = app_queue.get_static_pad("sink").unwrap();
+    tee_app_pad.link(&queue_app_pad).into_result().unwrap();
+
+    // configure appsrc
+    let info = AudioInfo::new(gst_audio::AudioFormat::S16le, SAMPLE_RATE, 1)
+        .build()
+        .unwrap();
+    let audio_caps = info.to_caps().unwrap();
+
+    let appsrc = appsrc.clone().dynamic_cast::<AppSrc>().expect(
+        "Source element is expected to be an appsrc!",
+    );
+    appsrc.set_caps(&audio_caps);
+    appsrc.set_property_format(gst::Format::Time);
+
+    let appsink = appsink.clone().dynamic_cast::<AppSink>().expect(
+        "Sink element is expected to be an appsink!",
+    );
+
+    let data: Arc<Mutex<CustomData>> =
+        Arc::new(Mutex::new(CustomData::new(appsrc.clone(), appsink.clone())));
+
+    let data_clone = data.clone();
+    appsrc.connect_need_data(move |_, _size| {
+        let data = &data_clone;
+        let mut d = data_clone.lock().unwrap();
+
+        if d.source_id.is_none() {
+            println!("start feeding");
+
+            let data_clone = data.clone();
+            d.source_id = Some(glib::source::idle_add(move || {
+                let data = &data_clone;
+                let num_samples = CHUNK_SIZE / 2; /* Each sample is 16 bits */
+                let mut buffer = gst::Buffer::with_size(CHUNK_SIZE).unwrap();
+
+                let appsrc = {
```

Comment:
`let (appsrc, buffer) = { ... }` would seem nicer, and you can also move the num_samples variable into the block